### PR TITLE
add design doc

### PR DESCRIPTION
**Pull request by skylian**
Let's move the design doc here.

### CODE REVIEW

**Review by yu239-zz — CHANGES_REQUESTED** [review]
Can you figure out how to enable syntax highlighting for quoted python code?

File: docs/design_doc.md [hunk 1]
```diff
@@ -0,0 +1,300 @@
+# PaddlePaddle Reinforcement Learning Framework (PARL)
+This is the design doc for [PARL](https://github.com/PaddlePaddle/PARL)：a general-purpose RL platform based on PaddlePaddle Fluid. 
+
+## Problem description
+
+> A robot is an intelligent entity that situates in an environment. At every time step, it receives multimodal sensory inputs and generates (possibly multimodal) action outputs according to a certain set of rewards or goals, given the current (partial) environment observation.
+
+Almost any RL problem can be seen in the above perspective. We can always convert a problem description to a one similar to the above (namely, where a robot takes actions in an environment at every time step and receives rewards).
+
+For example, in the scenario of learning to put only one advertise on each webpage, you can think of our decision-making algorithm as a robot, the user's online profile, browsing history, and some other contextual information as the environment observation, which advertisement to put on the webpage as the robot's action, and whether the user clicks on the advertisement as the reward. The time horizon in this case would then be a sequence of webpages.
```
Comment (emailweixu) [inline docs/design_doc.md]: advertise -> advertisement

File: docs/design_doc.md [hunk 2]
```diff
@@ -0,0 +1,300 @@
...
+## Representation of experience/sample
+An experience or a sample received by the robot is a struct that has four fields:
+- `inputs`: all the observed data at the current time step
+- `states`: memory data storing the previous history; useful for problems having temporal dependencies
+- `actions`: the actions taken by the robot at the time step
+- `rewards`: a list of scalar rewards for taking the actions
+- `game_over`: an integer value indicating the game-over status for the current time step
+Note that here the `game_over` field indicates the current time step instead of the next time step. This is different with most standard open-source simulator packages (see OpenAI gym for an example).
+#### Generalized actions: the concept of "control" no longer exists
+Different with our old RL framework, we now remove the concept of "control", and believe that anything that needs to be reinforced can be treated as a (generalized) action output, as long as the agent knows which actions really need to be sent back to the simulator.
```
Comment (emailweixu) [inline docs/design_doc.md]: Nobody knows our old RL framework

File: docs/design_doc.md [hunk 3]
```diff
@@ -0,0 +1,300 @@
...
+Note that here the `game_over` field indicates the current time step instead of the next time step. This is different with most standard open-source simulator packages (see OpenAI gym for an example).
+#### Generalized actions: the concept of "control" no longer exists
+Different with our old RL framework, we now remove the concept of "control", and believe that anything that needs to be reinforced can be treated as a (generalized) action output, as long as the agent knows which actions really need to be sent back to the simulator. 
+
+## Robot Components
+
+<center><img src="framework.png" width=900></center>
+A **Robot** consists of some **computation tasks**, i.e., applying some **algorithms** to using or learning **models**. For example, a robot has one reinforcement learning task and it uses an Actor-Critic algorithm to learn the policy model. To provide additional training of the policy model, the robot can also have some auxiliary task which is usually dealt with some unsupervised learning algorithm. Another scenario is where a robot needs different policies at different levels of temporal abstraction. In this case, we can assign one policy learning task to each policy.
```
Comment (emailweixu) [inline docs/design_doc.md]: which is usually dealt with => which usually deals with

File: docs/design_doc.md [hunk 4]
```diff
@@ -0,0 +1,300 @@
...
+
+In the following, we describe the design details of  the above-mentioned major components of robot. The the design principle is that we want to provide users with a framework that is easy for them to implement their own robots, by overriding or extending some of these components. The components of robot can be divided into two groups. One group handles the actual models, e.g., model computation, parameter sharing and synchronization; we call it **computation group**. The other group interacts with environment and produce/consume data to/from *computation group*; we call it **agent group** (we will explain this name later).
```
Comment (emailweixu) [inline docs/design_doc.md]: The the => The

Comment (emailweixu) [reply docs/design_doc.md]: The components of robot => The components of a robot

File: docs/design_doc.md [hunk 5]
```diff
@@ -0,0 +1,300 @@
...
+- There is no gradient flowing between two adjacent CTs.
+
+If you have two modules that must have gradients flow between them, then consider puting the two modules in one CT.
+Finally, different CTs might have different time resolutions. For example, a CT outputs actions at every time step, which another CT outputs actions very 100 time steps.
+
+<center><img src="ct.png"><img src="model.png"></center>
+
+<center>[**In the above figure, blue boxes represent pure Python logic control and computation, and purple boxes represent logic control and computation in Fluid programs.**](#)</center>
+
+A `ComputationTask` consists of two `Algorithm`s, a target algorithm and a reference algorithm. An algorithm should implement general prediction and learning functions, independent of the actual problem. The core of an algorithm is usually the back-end (optimization) of a network. An algorithm further contains a model which is problem-dependent and allows the user to write the front-end (data perception) of the network.
```
Comment (emailweixu) [inline docs/design_doc.md]: No need to have concepts of "target algorithm" and "reference algorithm". Target and/or reference should be details handled in algorithm, not in computation task

**Review by emailweixu — COMMENTED** [review]
Let's merge this doc first after fixing the minor problems. But clearly it isl not same as the final source code. So please change the doc accordingly in the PR for the actual code of the framework.

File: docs/design_doc.md [hunk 6]
```diff
@@ -0,0 +1,300 @@
...
+                                                                                 
+    def _prediction_loop(self):
+        while not stop:
+            agent_ids = []
+            data = []
+            while not agent_ids or not self.prediction_q.empty():
+                agent_id, d = self.prediction_q.get()
+                agent_ids.append(agent_id)
+                data.append(d)
+            data = self._pack_data(data)
```
Comment (TomorrowIsAnOtherDay) [inline docs/design_doc.md:194]: _unpack_data

File: docs/design_doc.md [hunk 7]
```diff
@@ -0,0 +1,300 @@
...
+            ret = self.ct.predict(data)
+            ret = self._unpack_data(ret)
```
Comment (TomorrowIsAnOtherDay) [inline docs/design_doc.md:196]: _pack_data

File: docs/design_doc.md [hunk 8]
```diff
@@ -0,0 +1,300 @@
...
+					self.helpers['RL'].learn()
+				obs = next_obs
+			self.helpers['goal'].store_data([obs, goal, total_reward, self.env.game_over()])
+			self.helpers['goal'].learn()
+```
+
+`ComputationWrapper` is essentially two threads that keep consuming data from Agent side:
+
+```python
+class ComputationWrapper(object):
```
Comment (TomorrowIsAnOtherDay) [inline docs/design_doc.md:159]: How does this Wrapper communicate with simulator?
Are there lots of simulators sending states into self.prediction_q asynchronously?

File: docs/design_doc.md (hunk 6 above)
Comment (skylian) [reply docs/design_doc.md:194]: it is the other way around. "pack" means putting multiple data from agents into one

File: docs/design_doc.md (hunk 8 above)
Comment (skylian) [reply docs/design_doc.md:159]: yes. multiple agents (processes) will send data into this q asynchronously

File: docs/design_doc.md [hunk 9]
```diff
@@ -64,7 +64,7 @@ Next, let's talk about the components in **agent group**: `Agent`, `AgentHelper`
 
 ### Agent Group
 
-<center><img src="relation.png" width=700></center>
+<p align="center"><img src="relation.png" width=600></p>
 <center>P_Q: prediction_q, PR_Q: prediction_return_q, T_Q: training_q, TR_Q: training_return_q</center>
```
Comment (yu239-zz) [inline docs/design_doc.md]: This line is not highlighted.

Comment (skylian) [reply docs/design_doc.md]: no need to highlight this line

**Review by yu239-zz — APPROVED** [review]

**Pull request merged**